crates/ra_ide_api/src/goto_definition.rs

```diff
@@ -123,9 +124,29 @@ pub(crate) fn reference_definition(
             Some(Resolution::SelfType(_impl_block)) => {
                 // TODO: go to the implemented type
             }
-            None => {}
+            None => {
+                // If we failed to resolve then check associated items
+                if let Some(function) = function {
+                    // Should we do this above and then grab path from the PathExpr?
```

Comment:
in a situation like `Foo::Bar<|>::Baz`, I think we ideally want to go to `Bar`, not `Baz`, and the above code actually handles that correctly; if you went up to the `PathExpr` and took the whole path from that, it wouldn't anymore.

To also handle it in the associated item case, we'd need to record the intermediate resolutions as well and store all of them in the map for the `ExprId` (maybe by index of the segment). Currently there are no such paths we can handle anyway, though. (I.e. if `Bar` is an associated item, it would have to be a type, and we don't deal with that yet.)

Reply:
I think I'm going to need more clarification on how to do this.

Reply:
I don't know if we actually should do this now, since this can't occur in the cases we can handle anyway. I'd be fine with leaving it like this for the moment.

(I think one way might be that we'd not just have a `FxHashMap<ExprId, ImplItem>`, but a `FxHashMap<(ExprId, usize), ImplItem>`, and then if we're on the `n`th path segment here (which we can see by looking at the length of the path we've got), we'd look up `(expr, n)` in the map.)

Reply:
Ok let's add it later then.